engine/installation/linux/ubuntulinux.md

```diff
@@ -116,365 +119,501 @@ packages from the Docker repository:
 
-Also, UFW's default set of rules denies all incoming traffic. If you want to
-reach your containers from another host allow incoming connections on the Docker
-port. The Docker port defaults to `2376` if TLS is enabled or `2375` when it is
-not. If TLS is not enabled, communication is unencrypted. By default, Docker
-runs without TLS enabled.
+In addition, UFW blocks all incoming traffic by default. If you want to access
+the Docker Remote API from another host and you have enabled remote access, you
+need to configure UFW to allow incoming connections on the Docker port, which
+defaults to `2376` if TLS encrypted transport is enabled or `2375` otherwise. By
+default, Docker runs **without** TLS enabled. If you do not use TLS, you are
+strongly discouraged from allowing access to the Docker Remote API from remote
+hosts, to prevent remote privilege-escalation attacks.
 
 To configure UFW and allow incoming connections on the Docker port:
 
-1. Log into Ubuntu as a user with `sudo` privileges.
+1.  Log into Ubuntu as a user with `sudo` privileges.
+
+2.  Verify that UFW is enabled.
 
-2. Verify that UFW is installed and enabled.
+    ```bash
+    $ sudo ufw status
+    ```
 
-        $ sudo ufw status
+    If `ufw` is not enabled, the remaining steps will not be helpful.
 
-3. Open the `/etc/default/ufw` file for editing.
+3.  If you need to enable access to the Docker Remote API from external hosts
+    and understand the security implications (see the section before this
+    procedure) edit the UFW configuration file, which is usually
+    `/etc/default/ufw` or `/etc/sysconfig/ufw`. Set the `DEFAULT_FORWARD_POLICY`
+    policy to `ACCEPT`.
 
-        $ sudo nano /etc/default/ufw
+    ```none
+    DEFAULT_FORWARD_POLICY="ACCEPT"
+    ```
 
-4. Set the `DEFAULT_FORWARD_POLICY` policy to:
+    Save and close the file.
 
-        DEFAULT_FORWARD_POLICY="ACCEPT"
+4.  Configure UFW to allow incoming connections on the Docker port, which is
+    2375 if you do not use TLS, and 2376 if you do.
 
-5. Save and close the file.
+    ```bash
+    $ sudo ufw allow 2375/tcp
+    ```
 
-6. Reload UFW to use the new setting.
+5.  Reload UFW.
+    ```bash
+    $ sudo ufw reload
+    ```
 
-        $ sudo ufw reload
+### Configure a DNS server for use by Docker
 
-7. Allow incoming connections on the Docker port.
+Ubuntu systems which use `networkmanager` use a `dnsmasq` instance that runs on
+a loopback address such as `127.0.0.1` or `127.0.1.1` and adds this entry to
+`/etc/resolv.conf`. The `dnsmasq` service provides a local DNS cache to speed up
+DNS look-ups and also provides DHCP services. However, Docker needs access to a
+fully-functional DNS server. If it detects that no DNS server referenced in
+`/etc/resolv.conf` is a fully functional DNS server, the following warning
+occurs and Docker uses the public DNS servers provided by Google at `8.8.8.8`
+and `8.8.4.4` for DNS resolution.
```

Comment:
fwiw; the main reason here is not "fully functional DNS", but the IP-address; the container has its own network namespace, and `127.x.x.x` inside the container is "localhost" _inside_ the container, not the host that the container runs on. Because of that, the container cannot connect to the DNS server, and will fail.

(The only exception is if the container is running with `--network=host`, because then it runs in the host's network namespace / network stack)